core/src/main/java/org/apache/carbondata/core/datastore/page/encoding/dimension/legacy/DictDimensionIndexCodec.java

```diff
@@ -46,18 +47,33 @@ public String getName() {
   public ColumnPageEncoder createEncoder(Map<String, String> parameter) {
     return new IndexStorageEncoder() {
       @Override
-      void encodeIndexStorage(ColumnPage inputPage) {
-        BlockIndexerStorage<byte[][]> indexStorage;
-        byte[][] data = inputPage.getByteArrayPage();
+      void encodeIndexStorage(ColumnPage input) {
+        BlockIndexerStorage<ByteBuffer[]> indexStorage;
+        boolean isDictionary = input.isLocalDictGeneratedPage();
+
+        // if need to build invertIndex or RLE, the columnpage should to be organized in Row,
+        // in the other words, we get the data of columnpage as an array, in which each element
+        // presenets a row. But if no need to build both invertIndex and RLE, it will increase
+        // extra overhead, considering data in columnpage was already stored as flattened data,
+        // and the compression is also on flattened  data, to organized data in ROW is actually
+        // increase the overheadof "Expand" and "Flatten" with on invertIndex and RLE.
+        // Overall, isFlatted presents do we flatten the data? if need to build invertIndex or RLE,
+        // isFlattened is set to ture, otherwise, isFlattened is set to false.
+        boolean isFlattened = !isInvertedIndex && !isDictionary;
+
+        // when isFlattened is true, data[0] is the flattened data of the columnpage.
+        // when isFlattened is false, data[i] is the ith row of the columnpage.
+        ByteBuffer[] data = input.getByteBufferArrayPage(isFlattened);
         if (isInvertedIndex) {
-          indexStorage = new BlockIndexerStorageForShort(data, true, false, isSort);
+          indexStorage = new BlockIndexerStorageForShort(data, isDictionary, !isDictionary, isSort);
```

Comment:
Not a good practice to pass same flag with complementary values. Hardcode is ok or introduce new variable